test/envtest/controllers/bootstrap_barrier_test.go

```diff
@@ -0,0 +1,226 @@
+					Racks: []scyllav1alpha1.RackNodesStatusReport{
+						{
+							Name: rackName,
+							Nodes: []scyllav1alpha1.NodeStatusReport{
+								{
+									Ordinal: 0,
+									HostID:  pointer.Ptr("host0"),
+									ObservedNodes: []scyllav1alpha1.ObservedNodeStatus{
+										{HostID: "host0", Status: scyllav1alpha1.NodeStatusUp},
+										{HostID: hostID, Status: scyllav1alpha1.NodeStatusUp},
+									},
+								},
+								{
+									Ordinal: nodeOrdinal,
+									HostID:  pointer.Ptr(hostID),
+									ObservedNodes: []scyllav1alpha1.ObservedNodeStatus{
+										{HostID: "host0", Status: scyllav1alpha1.NodeStatusUp},
+										{HostID: hostID, Status: scyllav1alpha1.NodeStatusUp},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			expectPreconditionMet: true,
+		},
+		{
+			name: "single status report with one node down",
+			objects: []client.Object{
+				testService(),
+				&scyllav1alpha1.ScyllaDBDatacenterNodesStatusReport{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "report-1",
+						Labels: map[string]string{
+							naming.ScyllaDBDatacenterNodesStatusReportSelectorLabel: nodesStatusReportSelector,
+						},
+					},
+					DatacenterName: dcName,
+					Racks: []scyllav1alpha1.RackNodesStatusReport{
+						{
+							Name: rackName,
+							Nodes: []scyllav1alpha1.NodeStatusReport{
+								{
+									Ordinal: 0,
+									HostID:  pointer.Ptr("host0"),
+									ObservedNodes: []scyllav1alpha1.ObservedNodeStatus{
+										{HostID: "host0", Status: scyllav1alpha1.NodeStatusUp},
+										{HostID: hostID, Status: scyllav1alpha1.NodeStatusDown},
+									},
+								},
+								{
+									Ordinal: nodeOrdinal,
+									HostID:  pointer.Ptr(hostID),
+									ObservedNodes: []scyllav1alpha1.ObservedNodeStatus{
+										{HostID: "host0", Status: scyllav1alpha1.NodeStatusUp},
+										{HostID: hostID, Status: scyllav1alpha1.NodeStatusDown},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			expectPreconditionMet: false,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			env := envtest.Setup(t)
+
+			t.Logf("Creating test objects")
+			for _, obj := range tc.objects {
+				if ok, err := env.KubeClient().IsObjectNamespaced(obj); err == nil && ok {
+					obj.SetNamespace(env.Namespace())
+				}
+
+				err := env.KubeClient().Create(t.Context(), obj)
+				if err != nil {
+					t.Fatalf("Failed to create object %T: %v", obj, err)
+				}
+				t.Logf("Created object %T/%s", obj, obj.GetName())
+			}
+
+			controllerOpts := controllerOpts{
+				nodeServiceName:                      nodeServiceName,
+				nodesStatusReportSelector:            nodesStatusReportSelector,
+				singleReportAllowNonReportingHostIDs: singleReportAllowNonReportingHostIDs,
+				bootstrapPreconditionSatisfiedCh:     make(chan struct{}),
+			}
+			runBoostrapBarrierController(t, env, controllerOpts)
+
+			if tc.expectPreconditionMet {
```

Comment:
can't the two blocks be merged? We could just compare the expected and got values, no?

Reply:
They're separate because the logic for both is slightly different:
- when `expectPreconditionMet=true`: wait until bootstrap precondition channel yields (success) or timeout (fatal)
- when `expectPreconditionMet=false`: wait 10s and expect channel to not yield (success); if it yields or the test context gets cancelled - fatal.